cou/apps/auxiliary.py

```diff
@@ -0,0 +1,165 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Auxiliary application class."""
+import logging
+from typing import Optional
+
+from cou.apps.app import AppFactory, OpenStackApplication
+from cou.exceptions import ApplicationError, HaltUpgradePlanGeneration
+from cou.steps import UpgradeStep
+from cou.utils.juju_utils import async_upgrade_charm
+from cou.utils.openstack import LTS_SERIES, OPENSTACK_TO_TRACK_MAPPING, OpenStackRelease
+
+logger = logging.getLogger(__name__)
+
+
+@AppFactory.register_application(["rabbitmq-server", "vault", "mysql", "ovn"])
+class OpenStackAuxiliaryApplication(OpenStackApplication):
+    """Application for charms that can have multiple OpenStack releases for a workload."""
+
+    def openstack_to_track(self, os_release: OpenStackRelease) -> str:
+        """Find the track of auxiliary charms by Ubuntu release and OpenStack release codename.
+
+        :param os_release: OpenStack release to track.
+        :type os_release: OpenStackRelease
+        :raises ApplicationError: When there is no track compatible.
+        :return: The track of the auxiliary charm.
+        :rtype: str
+        """
+        try:
+            return OPENSTACK_TO_TRACK_MAPPING[self.series][self.family][os_release.codename]
+        except KeyError as exc:
+            raise ApplicationError(
+                f"Cannot find a track of '{self.charm}' for {os_release.codename}"
+            ) from exc
+
+    def os_origin_config(self, target: OpenStackRelease) -> Optional[OpenStackRelease]:
+        """Identify the OpenStack release set on "openstack-origin" or "source" config.
+
+        :param target: OpenStack release as target to upgrade.
+        :type target: OpenStackRelease
+        :return: OpenStackRelease object or None if the app doesn't have os_origin config.
+        :rtype: Optional[OpenStackRelease]
+        """
+        # that means that the charm doesn't have "source" or "openstack-origin" config.
+        if self.origin_setting is None:
+            return None
+
+        # Ex: "cloud:focal-ussuri" will result in "ussuri"
+        os_origin_parsed: Optional[str] = self.os_origin.rsplit("-", maxsplit=1)[-1]
+        if os_origin_parsed == "distro":
+            # find the OpenStack release based on ubuntu series
+            os_origin_parsed = LTS_SERIES[self.series]
+        elif os_origin_parsed == "":
+            # if it's empty we consider the previous release from the target.
+            # Ex: rabbitmq-server has empty "source" and receive target "victoria".
+            # In that case it will be considered as ussuri and with the upgrade,
+            # "source" config will be changed to "cloud:focal-victoria".
+            os_origin_parsed = target.previous_release
```

Comment:
this is a big assumption... what if a user has configured a PPA?

Reply:
I think that for now it's ok, but we should provide warning to let user know that we are doing such assumption.

Reply:
In case user has configured ppa, this would raise an exception, because the string on the ppa won't match an valid OpenStack release codename. i'm handling this better in the new commit